cloud-controller-manager/do/certificates.go

```diff
@@ -16,8 +16,199 @@ limitations under the License.
 
 package do
 
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/digitalocean/godo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
 const (
 	// DO Certificate types
 	certTypeLetsEncrypt = "lets_encrypt"
 	certTypeCustom      = "custom"
+
+	// Certificate constants
+	certPrefix = "do-ccm-"
 )
+
+// ensureDomain checks to see if the service contains the annDODomain annotation
+// and if it does it verifies the domain exists on the users account
+func (l *loadBalancers) ensureDomain(ctx context.Context, service *v1.Service) error {
+	domain, err := getDomain(service)
+	if err != nil {
+		return err
+	}
+
+	if domain != nil {
+		klog.V(2).Infof("looking up root domain: %s", domain.root)
+		_, resp, err := l.resources.gclient.Domains.Get(ctx, domain.root)
+		if err != nil && resp.StatusCode == http.StatusNotFound {
+			return fmt.Errorf("domain does not exist: %s", domain.root)
+		} else if err != nil {
+			return fmt.Errorf("failed to retrieve domain: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// ensureCertificate verifies the existing certificate on the service is still a valid
+// DO certificate, clearing if not. If the user also has the annDODomain annotation set,
+// it verifies the certificate is valid for the domain, creating a new one if it cannot
+// find a existing valid certificate in the account
+func (l *loadBalancers) ensureCertificate(ctx context.Context, service *v1.Service) error {
```

Comment:
This method does a fair amount of things and feels non-trivial to follow to me.

Is it possible we split it up into two or three smaller pieces? I feel the the certificate lookup and (first) service update part would be a good scope for the first function (i.e., everything up to and including line 85). I'd return a certificate that the follow-up function (`ensureValidCertificateForDomain`?) would receive and process accordingly.

WDYT?

Reply:
Yeah completely agree. It will help lead to collapsing logic related to certificates in this function and `recordUpdatedLetsEncryptCert` https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/294/files#r346345188 as well.